lib/grppcaut.gi

```diff
@@ -869,12 +907,319 @@ end;
 ##
 #F AutomorphismGroupSolvableGroup( G )
 ##
+
+# should be more generic -- test whether orbit length exceeds limit
+BindGlobal("OrbitIsLonger",function(acts,pnt,act,lim)
+local orb,d,gen,i,p,D;
+  # try to find a domain
+  D:=DomainForAction(pnt,acts,act);
+  pnt:=Immutable(pnt);
+  d:=NewDictionary(pnt,false,D);
+  orb := [ pnt ];
+  AddDictionary(d,pnt);
+  for p in orb do
+    for gen in acts do
+      i:=act(p,gen);
+      MakeImmutable(i);
+      if not KnowsDictionary(d,i) then
+	Add( orb, i );
+	if Length(orb)>lim then
+	  return true;
+	fi;
+	AddDictionary(d,i);
+      fi;
+    od;
+  od;
+  return false;
+end);
+
+# another function that should be more general and has space for improvement.
+BindGlobal("SubspaceStabilizerMatrixGroup",function(G,bas)
+local f,act,xset,stb,hom,n,m,l,i;
+  n:=DimensionOfMatrixGroup(G);
+  if Length(bas)=0 or ForAll(GeneratorsOfGroup(G),
+	     x->ForAll(bas,y->SolutionMat(bas,y*x)<>fail)) then
+    # space is fixed
+    return G;
+  fi;
+  bas:=OnSubspacesByCanonicalBasis(bas,One(G)); # canonical form
```

Comment:
The reason I use `OnSubspacesByCanonicalBasis` is because we don't promise that the canonical form is exactly the result of `TriangulizeMat` (even though it currently is). As an action by `OnSubspaces...` follows using it this way avoids issues, even if the definition of one of the two operations ever changes.